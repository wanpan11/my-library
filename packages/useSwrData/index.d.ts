type SimpleKey = string | any[];
type SearchType<P> = P extends object ? Partial<P> : P;
interface PageType {
  pageNum: number;
  pageSize: number;
}

interface UseSwrDataFullProps<P = any, R = any> {
  reqKey: SimpleKey; // 请求的 key 可以是字符串或数组
  req: (params: P) => Promise<R>; // 请求函数 返回 Promise<AxiosRes<R>>
  ready?: boolean; // 是否准备就绪，默认为 true
  params?: SearchType<P>; // 受控请求参数 受控
  // paging 相关配置
  paging?: boolean; // 是否分页，默认为 false
  defaultPage?: PageType; // 默认分页信息 非受控
  defaultSearch?: SearchType<P>; // 默认搜索信息 非受控
  // swr 配置项
  swrConfig?: import("swr").SWRConfiguration<R, any>;
}
type UseSwrDataProps<P = any, R = any> = Omit<UseSwrDataFullProps<P, R>, "paging" | "defaultPage" | "defaultSearch">;

interface UseSwrData<R = any> {
  data?: R;
  error: any;
  isLoading: boolean;
  refresh: import("swr").KeyedMutator<R>;
}
interface UseSwrPagIngDataPage<P = any, R = any> extends UseSwrData<R> {
  pageInfo: PageType;
  searchInfo?: SearchType<P>;
  onSearch: (value: SearchType<P>) => void;
  setPage: React.Dispatch<React.SetStateAction<PageType>>;
  setSearch: React.Dispatch<React.SetStateAction<SearchType<P> | undefined>>;
}
