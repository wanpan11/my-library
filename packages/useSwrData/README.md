# use-swr-data

基于 [SWR](https://swr.vercel.app/) 封装的数据请求 Hook，提供了更便捷的数据获取和分页管理功能。

## 特性

- 🚀 简单易用的数据请求接口
- 📑 内置分页管理功能
- 🔍 支持搜索条件管理
- 💫 继承 SWR 的所有特性（缓存、自动重新验证等）
- 🎯 TypeScript 支持

## 安装

```bash
npm install u@wanp/use-swr-data
# 或
yarn add @wanp/use-swr-data
# 或
pnpm add @wanp/use-swr-data
```

## 基础使用

```typescript
import useSwrData from "@wanp/use-swr-data";

// 基础数据请求
const { data, error, isLoading, refresh } = useSwrData({
  reqKey: "unique-key",
  req: async (params) => {
    const response = await fetch("your-api-endpoint");
    return response.json();
  },
  params: {
    /* 可选的请求参数 */
  },
});
```

## 分页功能

```typescript
interface UserData {
  list: User[];
  total: number;
}

interface SearchParams {
  name?: string;
  age?: number;
}

const {
  data,
  error,
  isLoading,
  refresh,
  pageInfo,
  searchInfo,
  onSearch,
  setPage,
  setSearch,
} = useSwrData<UserData, SearchParams>({
  reqKey: "users",
  req: async (params) => {
    const response = await fetch(`/api/users?${new URLSearchParams(params)}`);
    return response.json();
  },
  paging: true,
  defaultPage: {
    pageNum: 1,
    pageSize: 10,
  },
  defaultSearch: {
    name: "",
    age: undefined,
  },
});
```

## API

### Props

#### 基础属性 (BaseSwrProps)

| 属性      | 类型                                  | 必填 | 描述                            |
| --------- | ------------------------------------- | ---- | ------------------------------- |
| reqKey    | `string \| string[]`                  | 是   | 请求的唯一标识                  |
| req       | `(params: TParams) => Promise<TData>` | 是   | 数据请求函数                    |
| ready     | `boolean`                             | 否   | 是否准备好发起请求，默认为 true |
| params    | `TParams \| Partial<TParams>`         | 否   | 受控请求参数                    |
| swrConfig | `SWRConfiguration`                    | 否   | SWR 配置项                      |

#### 分页属性 (PagingSwrProps)

继承基础属性，额外包含：

| 属性          | 类型               | 必填 | 描述         |
| ------------- | ------------------ | ---- | ------------ |
| paging        | `true`             | 是   | 启用分页功能 |
| defaultPage   | `PageInfo`         | 否   | 默认分页信息 |
| defaultSearch | `Partial<TParams>` | 否   | 默认搜索条件 |

### 返回值

#### 基础返回值 (BaseSwrResult)

| 属性      | 类型                  | 描述               |
| --------- | --------------------- | ------------------ |
| key       | `Key`                 | 当前请求的唯一标识 |
| data      | `TData`               | 请求返回的数据     |
| error     | `any`                 | 请求错误信息       |
| isLoading | `boolean`             | 是否正在加载       |
| refresh   | `KeyedMutator<TData>` | 手动刷新数据的方法 |

#### 分页返回值 (PagingSwrResult)

继承基础返回值，额外包含：

| 属性       | 类型                                 | 描述                         |
| ---------- | ------------------------------------ | ---------------------------- |
| pageInfo   | `PageInfo`                           | 当前分页信息                 |
| searchInfo | `Partial<TParams>`                   | 当前搜索条件                 |
| onSearch   | `(value: Partial<TParams>) => void`  | 搜索方法，会重置页码到第一页 |
| setPage    | `(page: PageInfo) => void`           | 更新分页信息                 |
| setSearch  | `(search: Partial<TParams>) => void` | 更新搜索条件                 |

## 注意事项

1. 分页模式下，`pageInfo` 和 `searchInfo` 的变更都会触发新的请求
2. `onSearch` 方法会自动重置页码到第一页
3. 默认禁用了 SWR 的 `revalidateOnFocus` 功能，可通过 `swrConfig` 开启
