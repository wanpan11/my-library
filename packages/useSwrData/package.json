{
  "name": "use-swr-data",
  "description": "",
  "version": "1.0.0",
  "type": "module",
  "main": "./dist/index.js",
  "types": "./dist/types/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "swr": "^2.3.4"
  }
}
