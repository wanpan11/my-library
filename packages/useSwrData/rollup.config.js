import typescript from "@rollup/plugin-typescript";

export default {
  input: "./src/index.ts",
  output: {
    dir: "dist",
    entryFileNames: "[name]-[format].js",
  },
  plugins: [
    typescript({
      compilerOptions: { declaration: true, declarationDir: "./dist/types" },
    }),
  ],
};
