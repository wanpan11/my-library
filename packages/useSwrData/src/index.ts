import type { Key } from "swr";
import useSwr from "swr";
import { useCallback, useMemo, useState } from "react";
import { DEFAULT_PAGE } from "./common";

function useSwrData<R = any, P = any>(props: UseSwrDataFullProps<P, R> & { paging: true }): UseSwrPagIngDataPage<P, R>;
function useSwrData<R = any, P = any>(props: UseSwrDataProps<P, R> & { paging?: false }): UseSwrData<R>;

function useSwrData<R = any, P = any>(props: UseSwrDataFullProps<P, R>): UseSwrData<R> | UseSwrPagIngDataPage<P, R> {
  const { reqKey, req, params, ready = true, paging = false, defaultSearch, defaultPage = DEFAULT_PAGE, swrConfig } = props;

  const [pageInfo, setPage] = useState(defaultPage);
  const [searchInfo, setSearch] = useState(defaultSearch);

  const mergeKey: Key = useMemo(() => {
    if (ready === false) {
      return null;
    }

    let mergeParams: any = {};
    if (paging) {
      mergeParams = { ...pageInfo, ...searchInfo, ...params };
    } else {
      mergeParams = params;
    }

    return [reqKey, mergeParams];
  }, [pageInfo, paging, params, ready, reqKey, searchInfo]);

  const { data, isLoading, error, mutate } = useSwr(
    mergeKey,
    async (data: [SimpleKey, P]) => {
      return req(data[1]);
    },
    swrConfig || { revalidateOnFocus: false },
  );

  const onSearch = useCallback(
    (value: SearchType<P>) => {
      setSearch(value);
      setPage(defaultPage);
    },
    [defaultPage],
  );

  if (paging) {
    return {
      data: data,
      error,
      isLoading,
      refresh: mutate,
      pageInfo,
      searchInfo,
      onSearch,
      setPage,
      setSearch,
    };
  } else {
    return {
      data: data,
      error,
      isLoading,
      refresh: mutate,
    };
  }
}

export default useSwrData;
