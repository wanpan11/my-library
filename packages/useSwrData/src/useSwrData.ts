import type { Key } from "swr";
import type { AnyObject, BaseSwrProps, BaseSwrResult, PagingSwrProps, PagingSwrResult, SimpleKey } from "./interface";
import { useCallback, useMemo, useState } from "react";
import useSwr from "swr";
import { DEFAULT_PAGE, SIMPLE_CONF } from "./common";

function useSwrData<TData = any, TParams = any>(props: BaseSwrProps<TData, TParams>): BaseSwrResult<TData>;
function useSwrData<TData = any, TParams extends AnyObject = any>(props: PagingSwrProps<TData, TParams>): PagingSwrResult<TData, TParams>;

function useSwrData<TData = any, TParams extends AnyObject = any>(
  props: BaseSwrProps<TData, TParams> | PagingSwrProps<TData, TParams>
): BaseSwrResult<TData> | PagingSwrResult<TData, TParams> {
  const { reqKey, req, params, ready = true, simple = false, paging = false, swrConfig } = props;

  const defaultPage = "defaultPage" in props ? props.defaultPage || DEFAULT_PAGE : DEFAULT_PAGE;
  const defaultSearch = "defaultSearch" in props ? props.defaultSearch : undefined;

  const [pageInfo, setPage] = useState(defaultPage);
  const [searchInfo, setSearch] = useState<Partial<TParams> | undefined>(defaultSearch);

  const mergeKey: Key = useMemo(() => {
    if (ready === false) {
      return null;
    }

    let mergeParams: any = {};
    if (paging) {
      mergeParams = { ...pageInfo, ...searchInfo, ...params };
    }
    else {
      mergeParams = params;
    }

    return [reqKey, mergeParams];
  }, [pageInfo, paging, params, ready, reqKey, searchInfo]);

  const mergeConf = useMemo(() => {
    return swrConfig || (simple ? SIMPLE_CONF : { revalidateOnFocus: false });
  }, [simple, swrConfig]);

  const { data, isLoading, error, mutate } = useSwr(
    mergeKey,
    async (data: [SimpleKey, TParams]) => {
      return req(data[1]);
    },
    mergeConf
  );

  const onSearch = useCallback(
    (value: Partial<TParams> | undefined) => {
      setSearch(value);
      setPage(defaultPage);
    },
    [defaultPage]
  );

  if (paging) {
    return {
      key: mergeKey,
      data,
      error,
      isLoading,
      refresh: mutate,
      pageInfo,
      searchInfo,
      onSearch,
      setPage,
      setSearch,
    };
  }
  else {
    return {
      key: mergeKey,
      data,
      error,
      isLoading,
      refresh: mutate,
    };
  }
}

export { useSwrData };
