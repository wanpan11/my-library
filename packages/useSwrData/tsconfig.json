{
  "compilerOptions": {
    "target": "ESNext",
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "baseUrl": ".",
    "declaration": true,
    "types": ["."]
  },
  "include": ["./src"]
}
