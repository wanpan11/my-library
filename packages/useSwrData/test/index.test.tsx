import { beforeEach, describe, expect, it, vi } from "vitest";
import { render } from "@testing-library/react";

import useSwrData from "../src/index";
import { afterEach } from "node:test";

type ObjParams = { name: string; age: number };
type NumParams = number;

const getReqParams = vi.fn(data => data);

function object(data: ObjParams) {
  return new Promise<ObjParams>(resolve => {
    setTimeout(() => {
      getReqParams(data);
      resolve(data);
    }, 1000);
  });
}
function num(data: NumParams) {
  return new Promise<NumParams>(resolve => {
    setTimeout(() => {
      getReqParams(data);
      resolve(data);
    }, 1000);
  });
}

function TestDemo<P>({ reqKey, req, params }: UseSwrDataProps<P>) {
  const { data, isLoading, error } = useSwrData({ reqKey, req, params });

  return (
    <div>
      <div>data: {JSON.stringify(data)}</div>
      <div>isLoading: {String(isLoading)}</div>
      <div>error: {String(error)}</div>
    </div>
  );
}
function TestDemoPaging<P>({ reqKey, req, params }: UseSwrDataFullProps<P>) {
  const { data, isLoading, error } = useSwrData({ reqKey, req, params, paging: true });

  return (
    <div>
      <div>data: {JSON.stringify(data)}</div>
      <div>isLoading: {String(isLoading)}</div>
      <div>error: {String(error)}</div>
    </div>
  );
}

describe("suite", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
    vi.restoreAllMocks();
  });

  it("obj params test", async () => {
    render(<TestDemo<ObjParams> reqKey={"obj-test"} req={object} params={{ name: "test", age: 18 }} />);

    await vi.advanceTimersByTimeAsync(2000);
    expect(getReqParams).toHaveReturnedWith({ name: "test", age: 18 });
  });

  it("num params test", async () => {
    render(<TestDemo<NumParams> reqKey={"num-test"} req={num} params={1} />);

    await vi.advanceTimersByTimeAsync(2000);
    expect(getReqParams).toHaveReturnedWith(1);
  });

  it("reqKey is array test", async () => {
    render(<TestDemo<NumParams> reqKey={["num-test", "2222"]} req={num} params={1} />);

    await vi.advanceTimersByTimeAsync(2000);
    expect(getReqParams).toHaveReturnedWith(1);
  });

  it("Paging test", async () => {
    render(<TestDemoPaging<NumParams> reqKey={["num-test", "2222"]} req={num} params={1} />);

    await vi.advanceTimersByTimeAsync(2000);
    expect(getReqParams).toHaveReturnedWith(1);
  });
});
