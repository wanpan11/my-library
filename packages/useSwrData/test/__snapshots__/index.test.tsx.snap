// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`suite > 1 1`] = `
<div>
  <div>
    <div>
      data: 
      {"name":"test","age":18}
    </div>
    <div>
      isLoading: 
      false
    </div>
    <div>
      error: 
      undefined
    </div>
  </div>
</div>
`;
